apps/src/blockly/addons/functionEditor.js

```diff
@@ -23,80 +111,237 @@ export default class FunctionEditor {
+      hiddenProcedure.getName(),
+      hiddenProcedure.getId()
+    );
+    this.editorWorkspace.getProcedureMap().add(editorProcedureModel);
+    Blockly.Events.enable();
+
+    this.showForFunction(hiddenProcedure);
+  }
+
+  handleDelete() {
+    // delete all caller blocks from the procedure workspace
+    Blockly.Procedures.getCallers(
+      this.block.getProcedureModel().getName(),
+      this.procedureWorkspace
+    ).forEach(block => {
+      block.dispose();
+    });
+
+    // delete all caller blocks from the main workspace
+    Blockly.Procedures.getCallers(
+      this.block.getProcedureModel().getName(),
+      this.mainWorkspace
+    ).forEach(block => {
+      block.dispose();
+    });
+
+    // delete the block from the editor workspace's procedure map
+    // this will also cause it to be deleted from the main and procedure
+    // workspaces' map
+    this.editorWorkspace
+      .getProcedureMap()
+      .delete(this.block.getProcedureModel().getId());
+
+    // delete the block from the editor workspace and hide the modal
+    this.block.dispose();
+    this.hide();
+  }
+
+  setUpEditorWorkspaceChangeListeners() {
+    // Mirror procedure events from editor workspace to main workspace.
+    // This allows updates for things like procedure name to propogate to the main
+    // workspace.
+    this.editorWorkspace.addChangeListener(e => {
+      // If the main workspace hasn't been initialized yet, don't do anything
+      if (!this.mainWorkspace) return;
+      if (e instanceof ProcedureBase) {
+        let event;
+        try {
+          event = Blockly.Events.fromJson(e.toJson(), this.mainWorkspace);
+        } catch (err) {
+          // Could not deserialize event. This is expected to happen. E.g. When
+          // round-tripping parameter deletes, the delete in the secondary workspace
+          // cannot be deserialized into the original workspace.
+          return;
+        }
+        event.run(true);
+
+        // Update the toolbox in case this change is happening
+        // while the flyout is open.
+        this.mainWorkspace.getToolbox().refreshSelection();
+      }
+    });
+
+    // Mirror all non-ui events from editor workspace to procedure workspace.
+    // This allows us to propogate edits to functions to the procedure workspace
+    // (the source of truth for function definitions).
+    this.editorWorkspace.addChangeListener(e => {
+      if (e.isUiEvent || !this.procedureWorkspace) return;
+      var json = e.toJson();
+      // Convert JSON back into an event, then execute it.
+      var secondaryEvent = Blockly.Events.fromJson(
+        json,
+        this.procedureWorkspace
+      );
+      secondaryEvent.run(true);
+    });
+  }
+
+  /**
+   * Add x and y coordinates to the block configuration so it shows up
+   * correctly in the editor workspace. When we copy the block between workspaces
+   * the x and y coordinates don't persist, so we need to add this to both new and
```

Comment:
Can you confirm deletable and moveable persist, even though coordinates don't?

Reply:
They do persist (I checked the block configs). However, I noticed I can still move the blocks, which I should put as a todo for follow-up. It's one of the many UI polish items I have tracked in my head 😬

Reply:
I made a [task](https://codedotorg.atlassian.net/browse/CT-7) for the UI polish items that were in my head